From d0a975e2480d17bf1a253ded32c0054b310434e3 Mon Sep 17 00:00:00 2001
From: Wang Hui <Hui.Wang@windriver.com>
Date: Fri, 6 Feb 2009 10:00:11 +0800
Subject: [PATCH] PowerPC/sbc8548:fix reboot failed

sbc8548 is unlike other members in mpc85xx family,
It cannot reset using the set-rstcr method, So I
change another way, that is set-dbcr0 method.

Signed-off-by: Wang Hui <Hui.Wang@windriver.com>
---
 arch/powerpc/platforms/85xx/sbc8548.c |   16 +++++++++++++++-
 1 files changed, 15 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/platforms/85xx/sbc8548.c b/arch/powerpc/platforms/85xx/sbc8548.c
index ecdd8c0..646174f 100644
--- a/arch/powerpc/platforms/85xx/sbc8548.c
+++ b/arch/powerpc/platforms/85xx/sbc8548.c
@@ -150,6 +150,20 @@ static void sbc8548_show_cpuinfo(struct seq_file *m)
 	seq_printf(m, "PLL setting\t: 0x%x\n", ((phid1 >> 24) & 0x3f));
 }
 
+static void sbc8548_restart(char* cmd)
+{
+	unsigned long msr, val;
+
+	msr = mfmsr();
+	msr |= MSR_DE;
+	mtmsr(msr);
+
+	val = mfspr(SPRN_DBCR0);
+	val |= 0x70000000;
+	mtspr(SPRN_DBCR0, val);
+	while(1);
+}
+
 static struct of_device_id __initdata of_bus_ids[] = {
 	{ .name = "soc", },
 	{ .type = "soc", },
@@ -183,7 +197,7 @@ define_machine(sbc8548) {
 	.init_IRQ	= sbc8548_pic_init,
 	.show_cpuinfo	= sbc8548_show_cpuinfo,
 	.get_irq	= mpic_get_irq,
-	.restart	= fsl_rstcr_restart,
+	.restart	= sbc8548_restart,
 #ifdef CONFIG_PCI
 	.pcibios_fixup_bus	= fsl_pcibios_fixup_bus,
 #endif
-- 
1.6.0.3

